### ISSUE TITLE
false negative based on filenames passed

### ISSUE THREAD

**Issue by asottile** [python/mypy#5760]
## version info

```console
$ pip freeze --all
mypy==0.630
mypy-extensions==0.4.1
pip==18.1
setuptools==40.4.3
typed-ast==1.1.0
wheel==0.32.1
```

## minimal reproduction

_note_ this was boiled down from a 13kloc codebase.  The unused imports in this example _are_ actually used in the real code.  The names are fake and the code is fake but the same issue is happening in reality in our code.

### `pkg/__init__.py`

```python
CONST = 'wat'
from . import bad  # fails as expected when this is removed
```

### `pkg/aaa.py`

```python
from typing import NamedTuple

from pkg import CONST  # fails as expected when removed


class NT(NamedTuple):  # fails as expected when changed to function
    x: str
```

### `pkg/bad.py`

```python
from typing import Optional

from pkg import CONST  # when removed, causes it to *always pass*
from pkg import aaa


def fallback() -> Optional[str]:
    return None


def f(x: Optional[str]) -> None:
    x = x or fallback()
    aaa.NT(x)  # should always be an error, NT requires `str`
```

## execution

Note, I expect all of these to fail in the same way.  There is a type error in the file but mypy is not picking it up.

Note that deleting the cache in-between each invocation is necessary as well, mypy seems to cache the bad state.

```console
$ rm -rf .mypy_cache/ && mypy pkg/aaa.py pkg/bad.py 
$ rm -rf .mypy_cache/ && mypy pkg/bad.py 
pkg/bad.py:13: error: Argument 1 to "NT" has incompatible type "Optional[str]"; expected "str"
$ rm -rf .mypy_cache/ && mypy pkg/bad.py pkg/aaa.py
pkg/bad.py:13: error: Argument 1 to "NT" has incompatible type "Optional[str]"; expected "str"
```

## revealing types

```diff
 def f(x: Optional[str]) -> None:
     x = x or fallback()
+    reveal_type(aaa.NT)
     aaa.NT(x)  # should always be an error, NT requires `str`
```

```console
$ rm -rf .mypy_cache/ && mypy pkg/aaa.py pkg/bad.py 
pkg/bad.py:13: error: Revealed type is 'Any'
$ rm -rf .mypy_cache/ && mypy pkg/bad.py 
pkg/bad.py:13: error: Revealed type is 'def (x: builtins.str) -> Tuple[builtins.str, fallback=pkg.aaa.NT]'
pkg/bad.py:14: error: Argument 1 to "NT" has incompatible type "Optional[str]"; expected "str"
```

**Comment by asottile** [comment]
this particular case appears to be fixed

```console
$ mypy --version
mypy 0.790
$ rm -rf .mypy_cache/ && mypy pkg/aaa.py pkg/bad.py 
pkg/bad.py:13: error: Argument 1 to "NT" has incompatible type "Optional[str]"; expected "str"
Found 1 error in 1 file (checked 2 source files)
$ rm -rf .mypy_cache/ && mypy pkg/bad.py 
pkg/bad.py:13: error: Argument 1 to "NT" has incompatible type "Optional[str]"; expected "str"
Found 1 error in 1 file (checked 1 source file)
$ rm -rf .mypy_cache/ && mypy pkg/bad.py pkg/aaa.py
pkg/bad.py:13: error: Argument 1 to "NT" has incompatible type "Optional[str]"; expected "str"
Found 1 error in 1 file (checked 2 source files)
```

looks like it was fixed in 0.720
